File: otelcollector/configmapparser/prometheus-ccp-config-merger.rb
#!/usr/local/bin/ruby
# frozen_string_literal: true

require "tomlrb"
require "deep_merge"
require "yaml"
require_relative "ConfigParseErrorLogger"

LOGGING_PREFIX = "prometheus-config-merger"
@configMapMountPath = "/etc/config/settings/prometheus/prometheus-config"
@promMergedConfigPath = "/opt/promMergedConfig.yml"
@mergedDefaultConfigPath = "/opt/defaultsMergedConfig.yml"
@replicasetControllerType = "replicaset"
@daemonsetControllerType = "daemonset"
@supportedSchemaVersion = true
@defaultPromConfigPathPrefix = "/opt/microsoft/otelcollector/default-prom-configs/"
@regexHashFile = "/opt/microsoft/configmapparser/config_def_targets_metrics_keep_list_hash"
@regexHash = {}
@sendDSUpMetric = false
@intervalHashFile = "/opt/microsoft/configmapparser/config_def_targets_scrape_intervals_hash"
@intervalHash = {}

@controlplane_apiserver_default_file = @defaultPromConfigPathPrefix + "controlplane_apiserver.yml"
@controlplane_kube_scheduler_default_file = @defaultPromConfigPathPrefix + "controlplane_kube_scheduler.yml"
@controlplane_kube_controller_manager_default_file = @defaultPromConfigPathPrefix + "controlplane_kube_controller_manager.yml"
@controlplane_cluster_autoscaler_default_file = @defaultPromConfigPathPrefix + "controlplane_cluster_autoscaler.yml"
@controlplane_etcd_default_file = @defaultPromConfigPathPrefix + "controlplane_etcd.yml"
@controlplane_prometheuscollectorhealth_default_file = @defaultPromConfigPathPrefix + "controlplane_prometheuscollectorhealth.yml"

def loadRegexHash
  begin
    @regexHash = YAML.load_file(@regexHashFile)
  rescue => errorStr
    ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Exception in loadRegexHash for prometheus config: #{errorStr}. Keep list regexes will not be used")
  end
end

def AppendMetricRelabelConfig(yamlConfigFile, keepListRegex)
  begin
    ConfigParseErrorLogger.log(LOGGING_PREFIX, "Adding keep list regex or minimal ingestion regex for #{yamlConfigFile}")
    config = YAML.load(File.read(yamlConfigFile))
    keepListMetricRelabelConfig = [{ "source_labels" => ["__name__"], "action" => "keep", "regex" => keepListRegex }]

    # Iterate through each scrape config and append metric relabel config for keep list
    if !config.nil?
      scrapeConfigs = config["scrape_configs"]
      if !scrapeConfigs.nil? && !scrapeConfigs.empty?
        scrapeConfigs.each { |scfg|
          metricRelabelCfgs = scfg["metric_relabel_configs"]
          if metricRelabelCfgs.nil?
            scfg["metric_relabel_configs"] = keepListMetricRelabelConfig
          else
            scfg["metric_relabel_configs"] = metricRelabelCfgs.concat(keepListMetricRelabelConfig)
          end
        }
        cfgYamlWithMetricRelabelConfig = YAML::dump(config)
        File.open(yamlConfigFile, "w") { |file| file.puts cfgYamlWithMetricRelabelConfig }
      end
    end
  rescue => errorStr
    ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Exception while appending metric relabel config in default target file - #{yamlConfigFile} : #{errorStr}. The keep list regex will not be used")
  end
end

# Get the list of default configs to be included in the otel's prometheus config
def populateDefaultPrometheusConfig
  begin
    # check if running in daemonset or replicaset
    currentControllerType = ENV["CONTROLLER_TYPE"].strip.downcase
    
    defaultConfigs = []
    if !ENV["AZMON_PROMETHEUS_CONTROLPLANE_KUBE_CONTROLLER_MANAGER_ENABLED"].nil? && ENV["AZMON_PROMETHEUS_CONTROLPLANE_KUBE_CONTROLLER_MANAGER_ENABLED"].downcase == "true" && currentControllerType == @replicasetControllerType
      kubeControllerManagerMetricsKeepListRegex = @regexHash["CONTROLPLANE_KUBE_CONTROLLER_MANAGER_KEEP_LIST_REGEX"]
      if !kubeControllerManagerMetricsKeepListRegex.nil? && !kubeControllerManagerMetricsKeepListRegex.empty?
        AppendMetricRelabelConfig(@controlplane_kube_controller_manager_default_file, kubeControllerManagerMetricsKeepListRegex)
      end
      contents = File.read(@controlplane_kube_controller_manager_default_file)
      contents = contents.gsub("$$POD_NAMESPACE$$", ENV["POD_NAMESPACE"])
      File.open(@controlplane_kube_controller_manager_default_file, "w") { |file| file.puts contents }
      defaultConfigs.push(@controlplane_kube_controller_manager_default_file)
    end

    if !ENV["AZMON_PROMETHEUS_CONTROLPLANE_KUBE_SCHEDULER_ENABLED"].nil? && ENV["AZMON_PROMETHEUS_CONTROLPLANE_KUBE_SCHEDULER_ENABLED"].downcase == "true" && currentControllerType == @replicasetControllerType
      controlplaneKubeSchedulerKeepListRegex = @regexHash["CONTROLPLANE_KUBE_SCHEDULER_KEEP_LIST_REGEX"]

      if !controlplaneKubeSchedulerKeepListRegex.nil? && !controlplaneKubeSchedulerKeepListRegex.empty?
        AppendMetricRelabelConfig(@controlplane_kube_scheduler_default_file, controlplaneKubeSchedulerKeepListRegex)
      end
      contents = File.read(@controlplane_kube_scheduler_default_file)
      contents = contents.gsub("$$POD_NAMESPACE$$", ENV["POD_NAMESPACE"])
      File.open(@controlplane_kube_scheduler_default_file, "w") { |file| file.puts contents }
      defaultConfigs.push(@controlplane_kube_scheduler_default_file)
    end

    if !ENV["AZMON_PROMETHEUS_CONTROLPLANE_APISERVER_ENABLED"].nil? && ENV["AZMON_PROMETHEUS_CONTROLPLANE_APISERVER_ENABLED"].downcase == "true" && currentControllerType == @replicasetControllerType
      controlplaneApiserverKeepListRegex = @regexHash["CONTROLPLANE_APISERVER_KEEP_LIST_REGEX"]
      
      if !controlplaneApiserverKeepListRegex.nil? && !controlplaneApiserverKeepListRegex.empty?
        AppendMetricRelabelConfig(@controlplane_apiserver_default_file, controlplaneApiserverKeepListRegex)
      end
      contents = File.read(@controlplane_apiserver_default_file)
      contents = contents.gsub("$$POD_NAMESPACE$$", ENV["POD_NAMESPACE"])
      File.open(@controlplane_apiserver_default_file, "w") { |file| file.puts contents }
      defaultConfigs.push(@controlplane_apiserver_default_file)
    end

    if !ENV["AZMON_PROMETHEUS_CONTROLPLANE_CLUSTER_AUTOSCALER_ENABLED"].nil? && ENV["AZMON_PROMETHEUS_CONTROLPLANE_CLUSTER_AUTOSCALER_ENABLED"].downcase == "true" && currentControllerType == @replicasetControllerType
      controlplaneClusterAutoscalerKeepListRegex = @regexHash["CONTROLPLANE_CLUSTER_AUTOSCALER_KEEP_LIST_REGEX"]
      
      if !controlplaneClusterAutoscalerKeepListRegex.nil? && !controlplaneClusterAutoscalerKeepListRegex.empty?
        AppendMetricRelabelConfig(@controlplane_cluster_autoscaler_default_file, controlplaneClusterAutoscalerKeepListRegex)
      end
      contents = File.read(@controlplane_cluster_autoscaler_default_file)
      contents = contents.gsub("$$POD_NAMESPACE$$", ENV["POD_NAMESPACE"])
      File.open(@controlplane_cluster_autoscaler_default_file, "w") { |file| file.puts contents }
      defaultConfigs.push(@controlplane_cluster_autoscaler_default_file)
    end

    if !ENV["AZMON_PROMETHEUS_CONTROLPLANE_ETCD_ENABLED"].nil? && ENV["AZMON_PROMETHEUS_CONTROLPLANE_ETCD_ENABLED"].downcase == "true" && currentControllerType == @replicasetControllerType
      controlplaneEtcdKeepListRegex = @regexHash["CONTROLPLANE_ETCD_KEEP_LIST_REGEX"]
      
      if !controlplaneEtcdKeepListRegex.nil? && !controlplaneEtcdKeepListRegex.empty?
        AppendMetricRelabelConfig(@controlplane_etcd_default_file, controlplaneEtcdKeepListRegex)
      end
      contents = File.read(@controlplane_etcd_default_file)
      contents = contents.gsub("$$POD_NAMESPACE$$", ENV["POD_NAMESPACE"])
      File.open(@controlplane_etcd_default_file, "w") { |file| file.puts contents }
      defaultConfigs.push(@controlplane_etcd_default_file)
    end

    if !ENV["AZMON_PROMETHEUS_NO_DEFAULT_SCRAPING_ENABLED"].nil? && ENV["AZMON_PROMETHEUS_NO_DEFAULT_SCRAPING_ENABLED"].downcase == "true"
      defaultConfigs.push(@controlplane_prometheuscollectorhealth_default_file)
    end

    @mergedDefaultConfigs = mergeDefaultScrapeConfigs(defaultConfigs)
  rescue => errorStr
    ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Exception while merging default scrape targets - #{errorStr}. No default scrape targets will be included")
    @mergedDefaultConfigs = ""
  end
end

def mergeDefaultScrapeConfigs(defaultScrapeConfigs)
  mergedDefaultConfigs = ""
  begin
    if defaultScrapeConfigs.length > 0
      mergedDefaultConfigs = YAML.load("scrape_configs:")
      # Load each of the default scrape configs and merge them
      defaultScrapeConfigs.each { |defaultScrapeConfig|
        # Load yaml from default config
        defaultConfigYaml = YAML.load(File.read(defaultScrapeConfig))
        mergedDefaultConfigs = mergedDefaultConfigs.deep_merge!(defaultConfigYaml)
      }
    end
    ConfigParseErrorLogger.log(LOGGING_PREFIX, "Done merging #{defaultScrapeConfigs.length} default prometheus config(s)")
  rescue => errorStr
    ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Exception while adding default scrape config- #{errorStr}. No default scrape targets will be included")
    mergedDefaultConfigs = ""
  end
  return mergedDefaultConfigs
end

def writeDefaultScrapeTargetsFile()
  ConfigParseErrorLogger.logSection(LOGGING_PREFIX, "Start Updating Default Prometheus Config")
  if !ENV["AZMON_PROMETHEUS_NO_DEFAULT_SCRAPING_ENABLED"].nil? && ENV["AZMON_PROMETHEUS_NO_DEFAULT_SCRAPING_ENABLED"].downcase == "false"
    begin
      loadRegexHash
      populateDefaultPrometheusConfig
      if !@mergedDefaultConfigs.nil? && !@mergedDefaultConfigs.empty?
        ConfigParseErrorLogger.log(LOGGING_PREFIX, "Starting to merge default prometheus config values in collector template as backup")
        mergedDefaultConfigYaml = YAML::dump(@mergedDefaultConfigs)
        File.open(@mergedDefaultConfigPath, "w") { |file| file.puts mergedDefaultConfigYaml }
      end
    rescue => errorStr
      ConfigParseErrorLogger.logError(LOGGING_PREFIX, "Error while populating default scrape targets and writing them to the default scrape targets file")
    end
  end
end

def setDefaultFileScrapeInterval(scrapeInterval)
  defaultFilesArray = [
    @controlplane_apiserver_default_file, @controlplane_kube_scheduler_default_file, @controlplane_kube_controller_manager_default_file,
    @controlplane_cluster_autoscaler_default_file, @controlplane_etcd_default_file, @controlplane_prometheuscollectorhealth_default_file
  ]

  defaultFilesArray.each { |currentFile|
    contents = File.read(currentFile)
    contents = contents.gsub("$$SCRAPE_INTERVAL$$", scrapeInterval)
    File.open(currentFile, "w") { |file| file.puts contents }
  }
end

setDefaultFileScrapeInterval("30s")
writeDefaultScrapeTargetsFile()

ConfigParseErrorLogger.logSection(LOGGING_PREFIX, "Done creating default targets file")

File: otelcollector/prometheuscollector/prometheus-ccp-config-merger.go
package main

import (
	"fmt"
	"io/fs"
	"os"
	"strings"

	"gopkg.in/yaml.v2"
)

const (
	mergedDefaultConfigPath               = "/opt/defaultsMergedConfig.yml"
	replicasetControllerType              = "replicaset"
	defaultPromConfigPathPrefix           = "/opt/microsoft/otelcollector/default-prom-configs/"
	regexHashFile                         = "/opt/microsoft/configmapparser/config_def_targets_metrics_keep_list_hash"
	controlplaneApiserverDefaultFile      = defaultPromConfigPathPrefix + "controlplane_apiserver.yml"
	controlplaneKubeSchedulerDefaultFile  = defaultPromConfigPathPrefix + "controlplane_kube_scheduler.yml"
	controlplaneKubeControllerManagerFile = defaultPromConfigPathPrefix + "controlplane_kube_controller_manager.yml"
	controlplaneClusterAutoscalerFile     = defaultPromConfigPathPrefix + "controlplane_cluster_autoscaler.yml"
	controlplaneEtcdDefaultFile           = defaultPromConfigPathPrefix + "controlplane_etcd.yml"
)

var (
	regexHash    = make(map[string]string)
	intervalHash = make(map[string]string)
)

var mergedDefaultConfigs map[interface{}]interface{}

func loadRegexHash() {
	data, err := os.ReadFile(regexHashFile)
	if err != nil {
		fmt.Printf("Exception in loadRegexHash for prometheus config: %v. Keep list regexes will not be used\n", err)
		return
	}

	err = yaml.Unmarshal(data, &regexHash)
	if err != nil {
		fmt.Printf("Exception in loadRegexHash for prometheus config: %v. Keep list regexes will not be used\n", err)
	}
}

func appendMetricRelabelConfig(yamlConfigFile, keepListRegex string) {
	fmt.Printf("Adding keep list regex or minimal ingestion regex for %s\n", yamlConfigFile)

	content, err := os.ReadFile(yamlConfigFile)
	if err != nil {
		fmt.Printf("Error reading config file %s: %v. The keep list regex will not be used\n", yamlConfigFile, err)
		return
	}

	var config map[string]interface{}
	if err := yaml.Unmarshal(content, &config); err != nil {
		fmt.Printf("Error unmarshalling YAML for %s: %v. The keep list regex will not be used\n", yamlConfigFile, err)
		return
	}

	keepListMetricRelabelConfig := map[string]interface{}{
		"source_labels": []interface{}{"__name__"},
		"action":        "keep",
		"regex":         keepListRegex,
	}

	if scrapeConfigs, ok := config["scrape_configs"].([]interface{}); ok {
		for _, scfg := range scrapeConfigs {
			if scfgMap, ok := scfg.(map[string]interface{}); ok {
				if metricRelabelCfgs, ok := scfgMap["metric_relabel_configs"].([]interface{}); ok {
					scfgMap["metric_relabel_configs"] = append(metricRelabelCfgs, keepListMetricRelabelConfig)
				} else {
					scfgMap["metric_relabel_configs"] = []interface{}{keepListMetricRelabelConfig}
				}
			}
		}

		if cfgYamlWithMetricRelabelConfig, err := yaml.Marshal(config); err == nil {
			if err := os.WriteFile(yamlConfigFile, []byte(cfgYamlWithMetricRelabelConfig), fs.FileMode(0644)); err != nil {
				fmt.Printf("Error writing to file %s: %v. The keep list regex will not be used\n", yamlConfigFile, err)
			}
		} else {
			fmt.Printf("Error marshalling YAML for %s: %v. The keep list regex will not be used\n", yamlConfigFile, err)
		}
	} else {
		fmt.Printf("No 'scrape_configs' found in the YAML. The keep list regex will not be used.\n")
	}
}

func populateDefaultPrometheusConfig() {
	loadRegexHash()

	defaultConfigs := []string{}
	currentControllerType := strings.TrimSpace(strings.ToLower(os.Getenv("CONTROLLER_TYPE")))

	if enabled, exists := os.LookupEnv("AZMON_PROMETHEUS_CONTROLPLANE_KUBE_CONTROLLER_MANAGER_ENABLED"); exists && strings.ToLower(enabled) == "true" && currentControllerType == replicasetControllerType {
		kubeControllerManagerMetricsKeepListRegex, exists := regexHash["CONTROLPLANE_KUBE_CONTROLLER_MANAGER_KEEP_LIST_REGEX"]
		if exists && kubeControllerManagerMetricsKeepListRegex != "" {
			appendMetricRelabelConfig(controlplaneKubeControllerManagerFile, kubeControllerManagerMetricsKeepListRegex)
		}
		contents, err := os.ReadFile(controlplaneKubeControllerManagerFile)
		if err == nil {
			contents = []byte(strings.Replace(string(contents), "$$POD_NAMESPACE$$", os.Getenv("POD_NAMESPACE"), -1))
			err = os.WriteFile(controlplaneKubeControllerManagerFile, contents, fs.FileMode(0644))
		}
		defaultConfigs = append(defaultConfigs, controlplaneKubeControllerManagerFile)
	}

	if enabled, exists := os.LookupEnv("AZMON_PROMETHEUS_CONTROLPLANE_KUBE_SCHEDULER_ENABLED"); exists && strings.ToLower(enabled) == "true" && currentControllerType == replicasetControllerType {
		controlplaneKubeSchedulerKeepListRegex, exists := regexHash["CONTROLPLANE_KUBE_SCHEDULER_KEEP_LIST_REGEX"]
		if exists && controlplaneKubeSchedulerKeepListRegex != "" {
			appendMetricRelabelConfig(controlplaneKubeSchedulerDefaultFile, controlplaneKubeSchedulerKeepListRegex)
		}
		contents, err := os.ReadFile(controlplaneKubeSchedulerDefaultFile)
		if err == nil {
			contents = []byte(strings.Replace(string(contents), "$$POD_NAMESPACE$$", os.Getenv("POD_NAMESPACE"), -1))
			err = os.WriteFile(controlplaneKubeSchedulerDefaultFile, contents, fs.FileMode(0644))
		}
		defaultConfigs = append(defaultConfigs, controlplaneKubeSchedulerDefaultFile)
	}

	if enabled, exists := os.LookupEnv("AZMON_PROMETHEUS_CONTROLPLANE_APISERVER_ENABLED"); exists && strings.ToLower(enabled) == "true" && currentControllerType == replicasetControllerType {
		controlplaneApiserverKeepListRegex, exists := regexHash["CONTROLPLANE_APISERVER_KEEP_LIST_REGEX"]
		if exists && controlplaneApiserverKeepListRegex != "" {
			appendMetricRelabelConfig(controlplaneApiserverDefaultFile, controlplaneApiserverKeepListRegex)
		}
		contents, err := os.ReadFile(controlplaneApiserverDefaultFile)
		if err == nil {
			contents = []byte(strings.Replace(string(contents), "$$POD_NAMESPACE$$", os.Getenv("POD_NAMESPACE"), -1))
			err = os.WriteFile(controlplaneApiserverDefaultFile, contents, fs.FileMode(0644))
		}
		defaultConfigs = append(defaultConfigs, controlplaneApiserverDefaultFile)
	}

	if enabled, exists := os.LookupEnv("AZMON_PROMETHEUS_CONTROLPLANE_CLUSTER_AUTOSCALER_ENABLED"); exists && strings.ToLower(enabled) == "true" && currentControllerType == replicasetControllerType {
		controlplaneClusterAutoscalerKeepListRegex, exists := regexHash["CONTROLPLANE_CLUSTER_AUTOSCALER_KEEP_LIST_REGEX"]
		if exists && controlplaneClusterAutoscalerKeepListRegex != "" {
			appendMetricRelabelConfig(controlplaneClusterAutoscalerFile, controlplaneClusterAutoscalerKeepListRegex)
		}
		contents, err := os.ReadFile(controlplaneClusterAutoscalerFile)
		if err == nil {
			contents = []byte(strings.Replace(string(contents), "$$POD_NAMESPACE$$", os.Getenv("POD_NAMESPACE"), -1))
			err = os.WriteFile(controlplaneClusterAutoscalerFile, contents, fs.FileMode(0644))
		}
		defaultConfigs = append(defaultConfigs, controlplaneClusterAutoscalerFile)
	}

	if enabled, exists := os.LookupEnv("AZMON_PROMETHEUS_CONTROLPLANE_ETCD_ENABLED"); exists && strings.ToLower(enabled) == "true" && currentControllerType == replicasetControllerType {
		controlplaneEtcdKeepListRegex, exists := regexHash["CONTROLPLANE_ETCD_KEEP_LIST_REGEX"]
		if exists && controlplaneEtcdKeepListRegex != "" {
			appendMetricRelabelConfig(controlplaneEtcdDefaultFile, controlplaneEtcdKeepListRegex)
		}
		contents, err := os.ReadFile(controlplaneEtcdDefaultFile)
		if err == nil {
			contents = []byte(strings.Replace(string(contents), "$$POD_NAMESPACE$$", os.Getenv("POD_NAMESPACE"), -1))
			err = os.WriteFile(controlplaneEtcdDefaultFile, contents, fs.FileMode(0644))
		}
		defaultConfigs = append(defaultConfigs, controlplaneEtcdDefaultFile)
	}

	mergedDefaultConfigs = mergeDefaultScrapeConfigs(defaultConfigs)
	if mergedDefaultConfigs != nil {
		fmt.Printf("Merged default scrape targets: %v\n", mergedDefaultConfigs)
	}
}

func mergeDefaultScrapeConfigs(defaultScrapeConfigs []string) map[interface{}]interface{} {
	for _, defaultScrapeConfig := range defaultScrapeConfigs {
		data, err := os.ReadFile(defaultScrapeConfig)
		if err != nil {
			fmt.Printf("Error reading default scrape config %s: %v. No default scrape targets will be included\n", defaultScrapeConfig, err)
			return nil
		}

		config := make(map[interface{}]interface{})
		err = yaml.Unmarshal(data, &config)
		if err != nil {
			fmt.Printf("Error unmarshalling YAML for default scrape config %s: %v. No default scrape targets will be included\n", defaultScrapeConfig, err)
			return nil
		}

		mergedDefaultConfigs = mergeYAML(mergedDefaultConfigs, config)
	}
	fmt.Printf("Done merging %d default prometheus config(s)\n", len(defaultScrapeConfigs))
	return mergedDefaultConfigs
}

func mergeYAML(dest, src map[interface{}]interface{}) map[interface{}]interface{} {
	for key, srcVal := range src {
		destVal, exists := dest[key]
		if exists {
			srcMap, isMap := srcVal.(map[interface{}]interface{})
			destMap, isDestMap := destVal.(map[interface{}]interface{})
			if isMap && isDestMap {
				dest[key] = mergeYAML(destMap, srcMap)
			} else {
				dest[key] = srcVal
			}
		} else {
			dest[key] = srcVal
		}
	}
	return dest
}

func writeDefaultScrapeTargetsFile() {
	fmt.Printf("Start Updating Default Prometheus Config\n")
	noDefaultScrapingEnabled := os.Getenv("AZMON_PROMETHEUS_NO_DEFAULT_SCRAPING_ENABLED")
	if noDefaultScrapingEnabled != "" && strings.ToLower(noDefaultScrapingEnabled) == "false" {
		loadRegexHash()
		populateDefaultPrometheusConfig()
		if mergedDefaultConfigs != nil && len(mergedDefaultConfigs) > 0 {
			fmt.Printf("Starting to merge default prometheus config values in collector template as backup\n")
			mergedDefaultConfigYaml, err := yaml.Marshal(mergedDefaultConfigs)
			if err == nil {
				err = os.WriteFile(mergedDefaultConfigPath, []byte(mergedDefaultConfigYaml), fs.FileMode(0644))
				if err != nil {
					fmt.Printf("Error writing merged default prometheus config to file: %v\n", err)
				}
			} else {
				fmt.Printf("Error marshalling merged default prometheus config: %v\n", err)
			}
		}
	} else {
		mergedDefaultConfigs = nil
	}
	fmt.Printf("Done creating default targets file\n")
}

func setDefaultFileScrapeInterval(scrapeInterval string) {
	defaultFilesArray := []string{
		controlplaneApiserverDefaultFile, controlplaneKubeSchedulerDefaultFile, controlplaneKubeControllerManagerFile,
		controlplaneClusterAutoscalerFile, controlplaneEtcdDefaultFile,
	}

	for _, currentFile := range defaultFilesArray {
		contents, err := os.ReadFile(currentFile)
		if err != nil {
			fmt.Printf("Error reading file %s: %v\n", currentFile, err)
			continue
		}

		contents = []byte(strings.Replace(string(contents), "$$SCRAPE_INTERVAL$$", scrapeInterval, -1))

		err = os.WriteFile(currentFile, contents, fs.FileMode(0644))
		if err != nil {
			fmt.Printf("Error writing to file %s: %v\n", currentFile, err)
		}
	}
}

func prometheusCcpConfigMerger() {
	mergedDefaultConfigs = make(map[interface{}]interface{}) // Initialize mergedDefaultConfigs
	setDefaultFileScrapeInterval("30s")
	writeDefaultScrapeTargetsFile()
	fmt.Printf("Done creating default targets file\n")
}
